oauth2_client/providers/oauth2/provider.py

```diff
@@ -0,0 +1,160 @@
+"""
+Abstract OAuth2 Provider class.
+
+To support a new OAuth2Provider you should extend this class. OAuth2Provider is also used for testing, to make sure
+some high-level features, like dynamic callback url registration, work fine.
+"""
+import logging
+from base64 import urlsafe_b64encode
+
+from oauthlib.oauth2 import BackendApplicationClient
+from requests_oauthlib import OAuth2Session
+
+from labsterutils.date_time import float_to_datetime, datetime_to_float
+from oauth2_client.compat import timestamp
+from oauth2_client.models import Application
+from oauth2_client.providers.base import Provider
+from django.utils import timezone
+from datetime import timedelta, datetime
+import pytz
+
+log = logging.getLogger(__name__)
+
+
+class OAuth2Provider(Provider):
+    id = 'oauth2'
+    name = 'Oauth2'
+
+    # JWT Bearer flow related constants
+    JWT_HEADER_RS256 = '{"alg":"RS256"}'
+    JWT_GRANT_TYPE = "urn:ietf:params:oauth:grant-type:jwt-bearer"
+
+    def fetch_token(self):
+        """
+        Top-level method, that returns the result of the wole auth flow - the access token
+        """
+        grant_to_fetcher = {
+            Application.GRANT_CLIENT_CREDENTIALS: self.fetch_token_client_credentials,
+            Application.GRANT_JWT_BEARER: self.fetch_token_jwt_bearer,
+        }
+        fetcher_f = grant_to_fetcher[self.app.authorization_grant_type]
+        raw_token = fetcher_f()
+        return self.accesstoken_from_raw_token(raw_token)
+
+    def fetch_token_client_credentials(self):
```

Comment:
[Optional] I'd rather create adapter for each grant type to avoid mixing logic in a single class.